### DISCUSSION TITLE
How to return custom status code (429) with React Server Components?

### DISCUSSION THREAD

**Discussion by moroshko** [vercel/next.js#50149] [App Router]
With `getServerSideProps`, when data is fetched and 429 (Too many requests) is returned, I do:
```
res.statusCode = 429;
res.end();
return { props: { error: "Too many requests" } };
```
How do I translate this response logic when fetching the data using React Server Components?

**Comment by proemian** [comment]
I don't think server components was meant to replicate traditional client/server semantics.
I only think you can accomplish this by calling an [api endpoint](https://nextjs.org/docs/app/building-your-application/routing/router-handlers) in `/app/api/..`, and return the errorcode using 
```return NextResponse.json({ error: 'Too many requests' }, { status: 429 });```

**Reply by moroshko** [reply]
I don't see how calling an API endpoint accomplishes that.
What I want is to return a 429 for a certain page that can't be rendered because of rate limiting.
@timneutkens @leerob @delbaoliveira What would you recommend?

**Comment by devjiwonchoi** [comment]
@moroshko Posted #53225 before working on it, can you confirm if this is what you are requesting?

**Reply by moroshko** [reply]
Commented there. Thanks for the initiative! @devjiwonchoi

**Comment by timneutkens** [comment]
Posted a reply here: https://github.com/vercel/next.js/discussions/53225#discussioncomment-6562537
